File: packages/core/src/engine.ts
/**
 * CKM engine — auto-generates topic index from a ckm.json manifest.
 *
 * @remarks
 * This module implements the SPEC.md algorithm for topic derivation,
 * JSON output, and terminal formatting. It handles both v1 and v2
 * manifests transparently.
 *
 * @packageDocumentation
 */

import { formatTopicContent, formatTopicIndex } from './format.js';
import { detectVersion, migrateV1toV2 } from './migrate.js';
import type {
  CkmConcept,
  CkmEngine,
  CkmErrorResult,
  CkmInspectResult,
  CkmManifest,
  CkmTopic,
  CkmTopicIndex,
  CkmTopicIndexEntry,
} from './types.js';

/**
 * Creates a CKM engine from a parsed manifest.
 *
 * @remarks
 * Main entry point. Accepts a parsed JSON object (v1 or v2).
 * If v1 is detected, auto-migrates to v2 internally.
 * Derives topics at construction time.
 * Returns a configured, immutable engine instance.
 *
 * @param data - Parsed CKM manifest (v1 or v2).
 * @returns A configured CKM engine.
 *
 * @public
 */
export function createCkmEngine(data: unknown): CkmEngine {
  const version = detectVersion(data);
  const manifest: CkmManifest =
    version === 1 ? migrateV1toV2(data as Record<string, unknown>) : (data as CkmManifest);

  const topics = deriveTopics(manifest);

  return {
    topics,
    getTopicIndex: (toolName = 'tool') => formatTopicIndex(topics, toolName),
    getTopicContent: (name) => formatTopicContent(topics, name),
    getTopicJson: (name) => {
      if (name === undefined || name === null) {
        return buildTopicIndexJson(topics, manifest);
      }
      return buildTopicJson(topics, name);
    },
    getManifest: () => manifest,
    inspect: () => buildInspect(manifest, topics),
  };
}

// ─── Topic Derivation (SPEC.md Section 3) ────────────────────────

function deriveTopics(manifest: CkmManifest): CkmTopic[] {
  const topics: CkmTopic[] = [];

  for (const concept of manifest.concepts) {
    // Step 1: Only concepts tagged "config" become topics
    if (!concept.tags.includes('config')) continue;

    const slug = concept.slug;
    if (!slug) continue;

    // Step 2: Collect related concepts
    const relatedConcepts: CkmConcept[] = [concept];
    for (const other of manifest.concepts) {
      if (other.id === concept.id) continue;
      const otherSlug = deriveSlugFromName(other.name);
      if (other.name.toLowerCase().includes(slug) || slug.includes(otherSlug)) {
        relatedConcepts.push(other);
      }
    }
    const conceptNames = relatedConcepts.map((c) => c.name);

    // Step 3: Match operations by tags or keywords
    const matchedOperations = manifest.operations.filter((op) => {
      if (hasTagOverlap(op.tags, [slug])) return true;
      return operationMatchesByKeyword(op, slug, conceptNames);
    });

    // Step 4: Match config entries by key prefix
    const matchedConfig = manifest.configSchema.filter((entry) => {
      const keyPrefix = entry.key.split('.')[0];
      return keyPrefix === slug;
    });

    // Step 5: Match constraints
    const matchedConstraints = manifest.constraints.filter((constraint) => {
      if (conceptNames.some((name) => constraint.enforcedBy.includes(name))) return true;
      return matchedOperations.some((op) => constraint.enforcedBy.includes(op.name));
    });

    // Step 6: Build topic
    topics.push({
      name: slug,
      summary: concept.what,
      concepts: relatedConcepts,
      operations: matchedOperations,
      configSchema: matchedConfig,
      constraints: matchedConstraints,
    });
  }

  return topics;
}

function deriveSlugFromName(name: string): string {
  let slug = name;
  if (slug.endsWith('Config')) slug = slug.slice(0, -6);
  else if (slug.endsWith('Result')) slug = slug.slice(0, -6);
  else if (slug.endsWith('Options')) slug = slug.slice(0, -7);
  return slug.toLowerCase();
}

function hasTagOverlap(tags1: readonly string[], tags2: readonly string[]): boolean {
  for (const t1 of tags1) {
    for (const t2 of tags2) {
      if (t1.toLowerCase() === t2.toLowerCase()) return true;
    }
  }
  return false;
}

function operationMatchesByKeyword(
  op: { name: string; what: string },
  slug: string,
  conceptNames: string[],
): boolean {
  const haystack = `${op.name} ${op.what}`.toLowerCase();
  if (haystack.includes(slug)) return true;
  return conceptNames.some((name) => haystack.includes(name.toLowerCase()));
}

// ─── JSON Output (SPEC.md Section 4.3, 4.4) ─────────────────────

function buildTopicIndexJson(topics: CkmTopic[], manifest: CkmManifest): CkmTopicIndex {
  return {
    topics: topics.map(
      (t): CkmTopicIndexEntry => ({
        name: t.name,
        summary: t.summary,
        concepts: t.concepts.length,
        operations: t.operations.length,
        configFields: t.configSchema.length,
        constraints: t.constraints.length,
      }),
    ),
    ckm: {
      concepts: manifest.concepts.length,
      operations: manifest.operations.length,
      constraints: manifest.constraints.length,
      workflows: manifest.workflows.length,
      configSchema: manifest.configSchema.length,
    },
  };
}

function buildTopicJson(topics: CkmTopic[], topicName: string): CkmTopic | CkmErrorResult {
  const topic = topics.find((t) => t.name === topicName);
  if (!topic) {
    return {
      error: `Unknown topic: ${topicName}`,
      topics: topics.map((t) => t.name),
    };
  }
  return topic;
}

// ─── Inspection (SPEC.md Section 5) ──────────────────────────────

function buildInspect(manifest: CkmManifest, topics: CkmTopic[]): CkmInspectResult {
  return {
    meta: manifest.meta,
    counts: {
      concepts: manifest.concepts.length,
      operations: manifest.operations.length,
      constraints: manifest.constraints.length,
      workflows: manifest.workflows.length,
      configKeys: manifest.configSchema.length,
      topics: topics.length,
    },
    topicNames: topics.map((t) => t.name),
  };
}

File: packages/rust-core/src/engine.rs
//! CKM engine — auto-generates topic index from a ckm.json manifest.
//!
//! Implements SPEC.md algorithms for topic derivation, JSON output, and
//! terminal formatting. Handles both v1 and v2 manifests transparently.

use serde_json::Value;

use crate::format::{format_topic_content, format_topic_index};
use crate::migrate::{derive_slug, detect_version, migrate_v1_to_v2};
use crate::types::{
    CkmConcept, CkmErrorResult, CkmInspectCounts, CkmInspectResult, CkmManifest, CkmManifestCounts,
    CkmTopic, CkmTopicIndex, CkmTopicIndexEntry, TopicJsonResult,
};

/// The core CKM engine.
///
/// Consumes a manifest (v1 or v2), derives topics at construction time,
/// and exposes methods for progressive disclosure at four levels.
#[derive(Debug, Clone)]
pub struct CkmEngine {
    manifest: CkmManifest,
    derived_topics: Vec<CkmTopic>,
}

impl CkmEngine {
    /// Creates a new CKM engine from a parsed JSON value.
    ///
    /// If the input is a v1 manifest, it is automatically migrated to v2.
    /// Topics are derived at construction time.
    pub fn new(data: Value) -> Self {
        let version = detect_version(&data);
        let manifest = if version == 1 {
            migrate_v1_to_v2(&data)
        } else {
            serde_json::from_value(data).unwrap_or_else(|_| CkmManifest {
                schema: String::new(),
                version: "2.0.0".to_string(),
                meta: crate::types::CkmMeta {
                    project: "unknown".to_string(),
                    language: "unknown".to_string(),
                    generator: "unknown".to_string(),
                    generated: String::new(),
                    source_url: None,
                },
                concepts: Vec::new(),
                operations: Vec::new(),
                constraints: Vec::new(),
                workflows: Vec::new(),
                config_schema: Vec::new(),
            })
        };

        let derived_topics = derive_topics(&manifest);

        CkmEngine {
            manifest,
            derived_topics,
        }
    }

    /// All auto-derived topics, computed at construction time.
    pub fn topics(&self) -> &[CkmTopic] {
        &self.derived_topics
    }

    /// Returns formatted topic index for terminal display (Level 0).
    ///
    /// Output stays within 300 token budget.
    pub fn topic_index(&self, tool_name: &str) -> String {
        format_topic_index(&self.derived_topics, tool_name)
    }

    /// Returns human-readable content for a specific topic (Level 1).
    ///
    /// Returns `None` if the topic is not found.
    /// Output stays within 800 token budget.
    pub fn topic_content(&self, topic_name: &str) -> Option<String> {
        format_topic_content(&self.derived_topics, topic_name)
    }

    /// Returns structured JSON data for a topic or the full index.
    ///
    /// - If `topic_name` is `None`: returns `TopicJsonResult::Index` (Level 2)
    /// - If `topic_name` matches a topic: returns `TopicJsonResult::Topic` (Level 1J)
    /// - If `topic_name` does not match: returns `TopicJsonResult::Error`
    pub fn topic_json(&self, topic_name: Option<&str>) -> TopicJsonResult {
        match topic_name {
            None => TopicJsonResult::Index(self.build_topic_index_json()),
            Some(name) => self.build_topic_json(name),
        }
    }

    /// Returns the raw manifest (v2, possibly migrated from v1).
    pub fn manifest(&self) -> &CkmManifest {
        &self.manifest
    }

    /// Returns manifest statistics: metadata, counts, and topic names.
    pub fn inspect(&self) -> CkmInspectResult {
        CkmInspectResult {
            meta: self.manifest.meta.clone(),
            counts: CkmInspectCounts {
                concepts: self.manifest.concepts.len(),
                operations: self.manifest.operations.len(),
                constraints: self.manifest.constraints.len(),
                workflows: self.manifest.workflows.len(),
                config_keys: self.manifest.config_schema.len(),
                topics: self.derived_topics.len(),
            },
            topic_names: self.derived_topics.iter().map(|t| t.name.clone()).collect(),
        }
    }

    fn build_topic_index_json(&self) -> CkmTopicIndex {
        CkmTopicIndex {
            topics: self
                .derived_topics
                .iter()
                .map(|t| CkmTopicIndexEntry {
                    name: t.name.clone(),
                    summary: t.summary.clone(),
                    concepts: t.concepts.len(),
                    operations: t.operations.len(),
                    config_fields: t.config_schema.len(),
                    constraints: t.constraints.len(),
                })
                .collect(),
            ckm: CkmManifestCounts {
                concepts: self.manifest.concepts.len(),
                operations: self.manifest.operations.len(),
                constraints: self.manifest.constraints.len(),
                workflows: self.manifest.workflows.len(),
                config_schema: self.manifest.config_schema.len(),
            },
        }
    }

    fn build_topic_json(&self, topic_name: &str) -> TopicJsonResult {
        match self.derived_topics.iter().find(|t| t.name == topic_name) {
            Some(topic) => TopicJsonResult::Topic(topic.clone()),
            None => TopicJsonResult::Error(CkmErrorResult {
                error: format!("Unknown topic: {}", topic_name),
                topics: self.derived_topics.iter().map(|t| t.name.clone()).collect(),
            }),
        }
    }
}

// ─── Topic Derivation (SPEC.md Section 3 — revised) ────────────────────
//
// Every concept with a non-empty slug becomes a topic.
// Operations/constraints/config are matched by tag overlap or keyword.
// Unmatched operations get their own topics so nothing is invisible.

fn derive_topics(manifest: &CkmManifest) -> Vec<CkmTopic> {
    let mut topics: Vec<CkmTopic> = Vec::new();
    let mut claimed_op_ids: Vec<String> = Vec::new();
    let mut claimed_constraint_ids: Vec<String> = Vec::new();

    // Phase 1: Every concept with a slug becomes a topic
    for concept in &manifest.concepts {
        let slug = &concept.slug;
        if slug.is_empty() {
            continue;
        }

        // Skip if we already have a topic with this slug (dedup)
        if topics.iter().any(|t| t.name == *slug) {
            // Merge this concept into the existing topic
            if let Some(existing) = topics.iter_mut().find(|t| t.name == *slug) {
                existing.concepts.push(concept.clone());
            }
            continue;
        }

        // Collect related concepts (same slug or name contains slug)
        let mut related_concepts: Vec<CkmConcept> = vec![concept.clone()];
        for other in &manifest.concepts {
            if other.id == concept.id {
                continue;
            }
            if other.slug == *slug {
                related_concepts.push(other.clone());
            } else {
                let other_slug = derive_slug(&other.name);
                if other.name.to_lowercase().contains(slug) || slug.contains(&other_slug) {
                    related_concepts.push(other.clone());
                }
            }
        }
        let concept_names: Vec<String> = related_concepts.iter().map(|c| c.name.clone()).collect();

        // Match operations by tag overlap or keyword
        let matched_operations: Vec<_> = manifest
            .operations
            .iter()
            .filter(|op| {
                if op.tags.iter().any(|t| t.to_lowercase() == slug.to_lowercase()) {
                    return true;
                }
                matches_by_keyword(&op.name, &op.what, slug, &concept_names)
            })
            .cloned()
            .collect();
        for op in &matched_operations {
            claimed_op_ids.push(op.id.clone());
        }

        // Match config entries by key prefix
        let matched_config: Vec<_> = manifest
            .config_schema
            .iter()
            .filter(|entry| {
                let key_prefix = entry.key.split('.').next().unwrap_or("");
                key_prefix == slug
            })
            .cloned()
            .collect();

        // Match constraints by enforcedBy referencing concepts or matched operations
        let matched_constraints: Vec<_> = manifest
            .constraints
            .iter()
            .filter(|c| {
                if concept_names.iter().any(|name| c.enforced_by.contains(name.as_str())) {
                    return true;
                }
                matched_operations.iter().any(|op| c.enforced_by.contains(op.name.as_str()))
            })
            .cloned()
            .collect();
        for c in &matched_constraints {
            claimed_constraint_ids.push(c.id.clone());
        }

        topics.push(CkmTopic {
            name: slug.clone(),
            summary: concept.what.clone(),
            concepts: related_concepts,
            operations: matched_operations,
            config_schema: matched_config,
            constraints: matched_constraints,
        });
    }

    // Phase 2: Unclaimed operations get their own topics
    for op in &manifest.operations {
        if claimed_op_ids.contains(&op.id) {
            continue;
        }
        let slug = derive_slug(&op.name);
        if slug.is_empty() {
            continue;
        }
        // If a topic with this slug already exists, add the operation to it
        if let Some(existing) = topics.iter_mut().find(|t| t.name == slug) {
            existing.operations.push(op.clone());
            claimed_op_ids.push(op.id.clone());
            continue;
        }
        // Create a new topic for this operation
        topics.push(CkmTopic {
            name: slug.clone(),
            summary: op.what.clone(),
            concepts: Vec::new(),
            operations: vec![op.clone()],
            config_schema: Vec::new(),
            constraints: Vec::new(),
        });
        claimed_op_ids.push(op.id.clone());
    }

    // Phase 3: Unclaimed constraints get added to matching topics or their own
    for constraint in &manifest.constraints {
        if claimed_constraint_ids.contains(&constraint.id) {
            continue;
        }
        // Try to find a topic whose operations match enforcedBy
        let mut matched = false;
        for topic in &mut topics {
            if topic.operations.iter().any(|op| constraint.enforced_by.contains(op.name.as_str())) {
                topic.constraints.push(constraint.clone());
                matched = true;
                break;
            }
        }
        if !matched {
            // Add to first topic or create an "other" topic
            if let Some(first) = topics.first_mut() {
                first.constraints.push(constraint.clone());
            }
        }
    }

    topics
}

/// Checks if a name+description matches a slug or concept names by keyword.
fn matches_by_keyword(name: &str, what: &str, slug: &str, concept_names: &[String]) -> bool {
    let haystack = format!("{} {}", name, what).to_lowercase();
    if haystack.contains(slug) {
        return true;
    }
    concept_names
        .iter()
        .any(|n| haystack.contains(&n.to_lowercase()))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn minimal_v2() -> Value {
        serde_json::json!({
            "$schema": "https://ckm.dev/schemas/v2.json",
            "version": "2.0.0",
            "meta": {
                "project": "test",
                "language": "typescript",
                "generator": "hand-authored",
                "generated": "2026-01-01T00:00:00.000Z"
            },
            "concepts": [
                {
                    "id": "concept-calver-config",
                    "name": "CalVerConfig",
                    "slug": "calver",
                    "what": "Configures CalVer validation rules.",
                    "tags": ["config"],
                    "properties": [
                        {
                            "name": "format",
                            "type": { "canonical": "string", "original": "CalVerFormat" },
                            "description": "Calendar format.",
                            "required": true,
                            "default": "YYYY.MM.DD"
                        }
                    ]
                }
            ],
            "operations": [
                {
                    "id": "op-validate",
                    "name": "validate",
                    "what": "Validates a calver version string.",
                    "tags": ["calver"],
                    "inputs": [
                        {
                            "name": "version",
                            "type": { "canonical": "string" },
                            "required": true,
                            "description": "The version string."
                        }
                    ]
                }
            ],
            "constraints": [
                {
                    "id": "constraint-no-future",
                    "rule": "No future dates.",
                    "enforcedBy": "validate",
                    "severity": "error"
                }
            ],
            "workflows": [],
            "configSchema": [
                {
                    "key": "calver.format",
                    "type": { "canonical": "string", "original": "CalVerFormat" },
                    "description": "Calendar format.",
                    "default": "YYYY.MM.DD",
                    "required": true
                }
            ]
        })
    }

    #[test]
    fn test_engine_new() {
        let engine = CkmEngine::new(minimal_v2());
        assert_eq!(engine.topics().len(), 1);
        assert_eq!(engine.topics()[0].name, "calver");
    }

    #[test]
    fn test_engine_topic_index() {
        let engine = CkmEngine::new(minimal_v2());
        let output = engine.topic_index("my-tool");
        assert!(output.contains("my-tool CKM"));
        assert!(output.contains("calver"));
    }

    #[test]
    fn test_engine_topic_content() {
        let engine = CkmEngine::new(minimal_v2());
        let output = engine.topic_content("calver").unwrap();
        assert!(output.contains("CalVerConfig"));
        assert!(engine.topic_content("nonexistent").is_none());
    }

    #[test]
    fn test_engine_topic_json_index() {
        let engine = CkmEngine::new(minimal_v2());
        match engine.topic_json(None) {
            TopicJsonResult::Index(idx) => {
                assert_eq!(idx.topics.len(), 1);
                assert_eq!(idx.topics[0].name, "calver");
            }
            _ => panic!("Expected TopicJsonResult::Index"),
        }
    }

    #[test]
    fn test_engine_topic_json_single() {
        let engine = CkmEngine::new(minimal_v2());
        match engine.topic_json(Some("calver")) {
            TopicJsonResult::Topic(t) => {
                assert_eq!(t.name, "calver");
            }
            _ => panic!("Expected TopicJsonResult::Topic"),
        }
    }

    #[test]
    fn test_engine_topic_json_error() {
        let engine = CkmEngine::new(minimal_v2());
        match engine.topic_json(Some("nonexistent")) {
            TopicJsonResult::Error(e) => {
                assert!(e.error.contains("Unknown topic"));
                assert!(e.topics.contains(&"calver".to_string()));
            }
            _ => panic!("Expected TopicJsonResult::Error"),
        }
    }

    #[test]
    fn test_engine_manifest() {
        let engine = CkmEngine::new(minimal_v2());
        assert_eq!(engine.manifest().meta.project, "test");
    }

    #[test]
    fn test_engine_inspect() {
        let engine = CkmEngine::new(minimal_v2());
        let result = engine.inspect();
        assert_eq!(result.meta.project, "test");
        assert_eq!(result.counts.concepts, 1);
        assert_eq!(result.counts.operations, 1);
        assert_eq!(result.counts.topics, 1);
        assert_eq!(result.topic_names, vec!["calver"]);
    }

    #[test]
    fn test_engine_v1_auto_migration() {
        let v1 = serde_json::json!({
            "project": "legacy",
            "generated": "2025-01-01T00:00:00.000Z",
            "concepts": [
                {
                    "id": "concept-CalVerConfig",
                    "name": "CalVerConfig",
                    "what": "Configures CalVer.",
                    "properties": [
                        {
                            "name": "format",
                            "type": "CalVerFormat",
                            "description": "The format."
                        }
                    ]
                }
            ],
            "operations": [],
            "constraints": [],
            "workflows": [],
            "configSchema": []
        });
        let engine = CkmEngine::new(v1);
        assert_eq!(engine.manifest().version, "2.0.0");
        assert_eq!(engine.manifest().meta.project, "legacy");
        assert_eq!(engine.topics().len(), 1);
        assert_eq!(engine.topics()[0].name, "calver");
    }

    #[test]
    fn test_engine_empty_manifest() {
        let data = serde_json::json!({
            "$schema": "https://ckm.dev/schemas/v2.json",
            "version": "2.0.0",
            "meta": {
                "project": "empty",
                "language": "rust",
                "generator": "hand-authored",
                "generated": "2026-01-01T00:00:00.000Z"
            },
            "concepts": [],
            "operations": [],
            "constraints": [],
            "workflows": [],
            "configSchema": []
        });
        let engine = CkmEngine::new(data);
        assert!(engine.topics().is_empty());
        assert_eq!(engine.inspect().counts.topics, 0);
    }
}
